### PR TITLE
Fix AspectJ basics

### PR DESCRIPTION
**Pull request by kriegaex**
## Fix basic AspectJ compiler functionality + add integration test

This is a first, basic step into improving AspectJ support for Plexus Compiler. It looks like nobody ever noticed, that target and release parameters are not forwarded to AJC, resulting in all class files being compiled to Java 1.2 target. I.e. that annotation-based @AspectJ syntax was completely unsupported before. Since Plexus was introduced 15 years ago, nobody ever did anything to upgrade the functionality a bit, only AspectJ versions were bumped before.

Furthermore, only `*.java` files were being considered for compilation, no `*.aj` files. This was also fixed, but at the cost of duplication with regard to redefining two static methods `getSourceFiles` and `getSourceFilesForSourceRoot`. Other compiler components did it similarly, so the necessary refactoring to make the methods non-static and break them down into smaller parts in order to be able to override only the parts of them dealing with source file extensions, is a TODO which involves refactoring the other compiler components, too. I did not do that in this first step.

Another TODO: In contrast to the ECJ adapter, which uses the batch compiler, AJC (an ECJ fork!) is used via the internal AJDT interface, which is designed to be used by the Eclipse IDE. It would have been easier to actually also use the batch compiler right from the start, because then we can more easily map command line parameters and keep the adapter layer thin. Why the original author did that in 2005, remains a mystery.

The new AspectJ integration test uses a mixture of
  - native and @AspectJ syntax variants,
  - `*.java` and `*.aj` aspect source files,
  - production (`src/main`) and test (`src/test`) aspects.

This is all covered by a single IT. In the end, the test asserts on the complete JUnit console log, which has to look a certain way in order to reflect that
  - compilation and test were successful,
  - all 3 aspects did what they are supposed to do.

### CODE REVIEW

**Comment by kriegaex** [comment]
@olamy, like we said in the e-mail exchange: I am going to look into AspectJ compilation via Plexus and you can review and merge, if applicable. A lot more needs to be done here, but after looking into this today for the first time, at least I wanted to enable basic AspectJ functionality for the first time in 15+ years. The previous base class unit test only tested AJC as a drop-in replacement for Javac, no actual AspectJ.

**Comment by kriegaex** [comment]
The IT failing is the one for the ErrorProne integration on some JDK-17-EA configurations. It has been failing since shortly after 17-EA was added to the build matrix. The new AspectJ IT and the other tests are passing. So I guess, this problem should be fixed on master:

image

File: plexus-compiler-its/pom.xml [hunk 1]
```diff
@@ -39,6 +39,8 @@
               <goal>verify</goal>
             </goals>
             <configuration>
+              <!-- If you want to run a single IT, use 'invokerTest' -->
```
Comment (olamy) [inline plexus-compiler-its/pom.xml:42]: please remove, The best option is to use cli with `-Dinvoker.test=`

File: plexus-compiler-its/src/main/it/aspectj-compiler/src/main/java/org/acme/do-not-try-to-compile.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+The Maven Compiler (MC) API only allows a single source file extension, such as '.java'.
```
Comment (olamy) [inline plexus-compiler-its/src/main/it/aspectj-compiler/src/main/java/org/acme/do-not-try-to-compile.txt:1]:
```suggestion
The Apache Maven Compiler (MC) API only allows a single source file extension, such as '.java'.
```

**Review by olamy — CHANGES_REQUESTED** [review]
thanks for work here. only very nitpick comments.

File: plexus-compiler-its/pom.xml (hunk 1 above)
Comment (kriegaex) [reply plexus-compiler-its/pom.xml:42]: I can remove it, but I put it there on purpose, because like many other developers, usually I just run Maven builds by clicking on goals from an IDE, being able to also click on errors there and get to the problematic code quickly, instead of fabricating a whole command line outside the IDE, manually selecting profiles etc. I left this hint behind for other developers who want to do the same in a convenient way, because while most people know `-DskipTests=true` and `-Dtest=MyTest`, many of them are not Maven Invoker gurus (myself included) and don't want to google the property name if once in 3 months an IT fails and they cannot remember. Please do not assume, that your level of deep Maven CLI knowledge is the same as everybody else's. It is just a comment and I left it there for the convenience of others as much as myself.

File: plexus-compiler-its/src/main/it/aspectj-compiler/src/main/java/org/acme/do-not-try-to-compile.txt (hunk 2 above)
Comment (kriegaex) [reply plexus-compiler-its/src/main/it/aspectj-compiler/src/main/java/org/acme/do-not-try-to-compile.txt:1]: Why is the branding necessary here?

Comment (olamy) [reply plexus-compiler-its/src/main/it/aspectj-compiler/src/main/java/org/acme/do-not-try-to-compile.txt:1]: as long as you mention Maven it should Apache Maven. That's Apache branding requirement for any Apache product

File: plexus-compiler-its/pom.xml (hunk 1 above)
Comment (olamy) [reply plexus-compiler-its/pom.xml:42]: by experience I know people will have such piece of code, forget to put it back within comments then commit/push.

Comment (bmarwell) [reply plexus-compiler-its/pom.xml:42]: IT is CLI first. If the IDE cannot cope with it, I open a bug at jetbrains/eclipse.org etc.
If you run a maven goal from any IDE, you can add `-Dinvoker.test=testname` there, too.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java [hunk 3]
```diff
@@ -565,75 +572,164 @@ private void checkForAspectJRT( List<String> cp )
...
+        VERSION_MAP.put( "12",   JDK12 );
+        VERSION_MAP.put( "12.0", JDK12 );
+        VERSION_MAP.put( "13",   JDK13 );
+        VERSION_MAP.put( "13.0", JDK13 );
+        VERSION_MAP.put( "14",   JDK14 );
+        VERSION_MAP.put( "14.0", JDK14 );
+        VERSION_MAP.put( "15",   JDK15 );
+        VERSION_MAP.put( "15.0", JDK15 );
+        VERSION_MAP.put( "16",   JDK16 );
+        VERSION_MAP.put( "16.0", JDK16 );
```
Comment (bmarwell) [inline plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:629]: Haven't checked, but I think there is a JavaVersion class somewhere in the class path.
Also, do we need the `.0` versions starting from 9? I never encountered them.
And while at it, you can remove 5, 5.0, 6, 6.0, 7, 7.0.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java [hunk 4]
```diff
@@ -565,75 +572,164 @@ private void checkForAspectJRT( List<String> cp )
...
+
+    static
+    {
+        // Default in Maven Compiler 3.8.1
+        VERSION_MAP.put( "",     JDK1_6 );
+
+        VERSION_MAP.put( "1.1",  JDK1_1 );
+        VERSION_MAP.put( "1.2",  JDK1_2 );
+        VERSION_MAP.put( "1.3",  JDK1_3 );
+        VERSION_MAP.put( "1.4",  JDK1_4 );
```
Comment (bmarwell) [inline plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:600]: Did aspectj ever support these versions? If not, it is probably safe to remove them.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java [hunk 5]
```diff
@@ -25,9 +25,14 @@
...
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+
+import static org.aspectj.org.eclipse.jdt.internal.compiler.classfmt.ClassFileConstants.*;
```
Comment (elharo) [inline plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:35]: I'd prefer not to static import these as they're not familiar and common so readers will wonder where they came from.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java [hunk 6]
```diff
@@ -565,75 +572,164 @@ private void checkForAspectJRT( List<String> cp )
...
-        else if ( "1.8".equals( sourceVersion ) )
-        {
-            buildConfig.getOptions().sourceLevel = ClassFileConstants.JDK1_8;
-        }
-        else if ( "1.7".equals( sourceVersion ) )
+        buildConfig.getOptions().sourceLevel = versionStringToMajorMinor( sourceVersion );
+    }
+
+    /**
+     * Set the target version in ajc compiler
```
Comment (elharo) [inline plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:579]: nit: period and AJC

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 3 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:629]: > Also, do we need the .0 versions starting from 9? I never encountered them.

AspectJ is an ECJ fork, and [ECJ supports those version numbers](https://github.com/eclipse/eclipse.jdt.core/blob/6c5346e6b20e0e2cfc2fc914b5d2eaa9194e3914/org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/messages.properties#L239-L258), hence AspectJ does too.

Actually, this can go away in the future, if I switch from the internal AJDT compiler interface to AJC being a subclass of ECJ, merely extending instead of redefining it.

> And while at it, you can remove 5, 5.0, 6, 6.0, 7, 7.0.

Those are legal for ECJ/AJC too, see above.

BTW, that verbose way to create a map is ugly - still better than the if-else (or was it swirtch-case) from before, but if on the class path there is a tool class somewhere, permitting me to do this in a more compact way than using lots of `put` statements, please let me know. I just wanted to get it working and be reusable for all of source, target and release parameters in a uniform way. Before, only source was covered.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 4 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:600]: Yes, it does because ACJ does. Remember, AspectJ is also a drop-in replacement for simple Java files without any aspects. This even is what happened before, because the target parameter was not passed on: AJC always created Java 1.2 class files, just like I said in the first sentence of this PR.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 5 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:35]: Sorry, but I use so many of them, it is just too wordy otherwise with all Java versions from 1.1 to 16. You would not see the forest because of the many trees anymore. And I expect other readers (who are always developers) to either understand what a static import is - or let their IDE help them when they jump to a definition.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 6 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:579]: 1. I merely copied the Javadoc from above, so this is not my choice of spelling.
2. All of ajc, Ajc and AJC are in use. Actually, AJC is rather rare among us AspectJ developers. Sometimes we use ajc, because that is the name of the executable. Sometimes we use Ajc when comparing it to Javac as a proper name in prose. Actually, AJC is rather synthetic, in order to look more similar to ECJ when comparing those.

There is no right or wrong here. I copied it, I leave it at that. Otherwise I have to change existing comments in the whole module too.

File: plexus-compiler-its/pom.xml (hunk 1 above)
Comment (kriegaex) [reply plexus-compiler-its/pom.xml:42]: Fine, as you wish. I am going to remove it. *shrug*

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 6 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:579]: I am simply going to call in "AspectJ compiler", because "AJC compiler" actually means "AspectJ compiler compiler" and is redundant. It is just used in two comments, the one I copied and this one. I am not going to add the period, though, because it is a single sentence, basically like a subject line. The existing javadocs are so bad, I am not going to start revising all of them. For 15 years nobody even so much as looked at this code, and now the code review is about comments in POMs and Javadocs.

File: plexus-compiler-its/src/main/it/aspectj-compiler/src/main/java/org/acme/do-not-try-to-compile.txt (hunk 2 above)
Comment (kriegaex) [reply plexus-compiler-its/src/main/it/aspectj-compiler/src/main/java/org/acme/do-not-try-to-compile.txt:1]: OMG. Fine. This is exactly the kind of "code" review I was afraid of, when we first talked in the issue over atmy AspectJ Maven fork - sorry, AspectJ _Apache_ Maven fork.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 5 above)
Comment (elharo) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:35]: You can't rely on IDEs here. For instance, I am currently reading this code in my web browser, not an IDE.

Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:35]: I said "or IDE". You don't need to rely on it in order to understand the code. I hope you don't mind that I am not going to cater to your personal preference of 19 static imports (I counted!), because you think it makes the code more readable. I simply disagree.

**Comment by kriegaex** [comment]
Can we please conclude this review? I am waiting for feedback. I think I have addressed all comments with corresponding commits or explanations. Is there anything else I need to do, or are you guys just busy?

**Comment by bmarwell** [comment]
Hi Alex, for me the code style changes are still open.
I really can see your point why you want to static import constants (I would do the same in my project), but I think keeping the same style is even more important than that.
Besides, you IDE will help you anyway.
I suggest just importing the class `import x.y..ClassFileConstants` and then referencing with `ClassFileConstants.x`. 
But I will see if I can get someone else to look over it, so we might get an exception. I feel this could be justified.

Thanks for the patch anyway, you got my +1 (apart from the open nits) for its contents!

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 5 above)
Comment (rfscholte) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:35]: Codehaus style is explicit imports, consistency is key over personal preferences.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java [hunk 7]
```diff
@@ -295,7 +300,7 @@
 
     public AspectJCompiler()
     {
-        super( CompilerOutputStyle.ONE_OUTPUT_FILE_PER_INPUT_FILE, ".java", ".class", null );
+        super( CompilerOutputStyle.ONE_OUTPUT_FILE_PER_INPUT_FILE, "", ".class", null );
```
Comment (rfscholte) [inline plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:303]: I assume this is a hack to support `.aj` and `.java`. Should be added as a comment, because without that context this line doesn't make any sense.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 3 above)
Comment (rfscholte) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:629]: org.codehaus.plexus.languages.java.version.JavaVersion is the solution to reduce this list, which will turn it back into if/switch statements :)

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java [hunk 8]
```diff
@@ -557,83 +564,172 @@ private void checkForAspectJRT( List<String> cp )
...
+    {
+        buildConfig.getOptions().targetJDK = versionStringToMajorMinor( targetVersion );
+    }
+
+    private static final Map<String, Long> VERSION_MAP = new HashMap<>();
+
+    static
+    {
+        // Default in Maven Compiler 3.8.1
+        VERSION_MAP.put( "",     JDK1_6 );
```
Comment (rfscholte) [inline plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:595]: This looks like a weird assumption to me: a library that assumes it is being called by a specific version of a plugin. Should be removed.

**Review by rfscholte — CHANGES_REQUESTED** [review]

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 8 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:595]: The comment only explains why I chose this default for an empty string in 2021. The version number is a hint about the origin of this choice. I can choose anything, but I wanted to document why I made the choice, so that in the future someone else can update the default, based on another up-to-date Maven Compiler version. Any default is OK, if it is documented. That is to yet be done, of course, because the Plexus components are lacking meaningful Maven sites.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 7 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:303]: What you call a "hack" is the only way to achieve the desired behaviour. I am just doing what Groovy-Eclipse does, too. The alternative would be to extens Maven Compiler to support a set of source file extensions. I even briefly looked into that, but it would have meant to
  * add an opverloaded constructor,
  * keep the getter for the single value for backward compatibility,
  * add another getter for a set of values,
  * handling the case that the single extension getter is called in case multiple extensions exist (throw an exception? return the first value? what is the first value in a set? do we need a unique list instead?).

I wanted to keep it simple and did this. See lines 663 through 668.

I am going to add a source code comment. I just wanted to get the "hack" argument off the table first. 🙂

**Review by olamy — APPROVED** [review]
Thanks for the contribution Good job  (sounds good to me maybe except the import static I don't like much)

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 5 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:35]: Oh OK, I did not know that. Fine, I am going to change it.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 3 above)
Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:629]: `plexus-languages` is not on the classpath. Is it worth pulling it in for this?

Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:629]: OK, I can answer my own question already: It does not make sense. First of all, it cannot parse `8.0` correctly into a major `8` (still returns major `8.0`), but ECJ/AJC support it. Also comparing versions does not work correctly (e.g. `JavaVersion.parse("8").isBefore("1.9")` returns `false)`. Furthermore, there would zero benefit, even if that utility worked perfectly, because I need to convert strings into AJDT `ClassFileConstants.*` values of type `long`, encoding major-minor in a way not provided by the Plexus `JavaVersion`, e.g. `ClassFileConstants.JDK1_8` evaluates to `3407872`.

so I think we can scratch this and leave it as it is for now. Like I already explained, if in the future any developer decides to switch from the ADJT interfact to the AJC batch compiler, we can let the compiler do the parsing. But the AJDT API is more strongly typed, not dependent on only parsing strings from a CLI. So for now this is how it is.

**Comment by kriegaex** [comment]
Some general remarks:
* First of all, thanks for all the reviews, even though I am kind of irritated by the code review style here. It is not how I teach my agile teams to do code reviews, but I am trying to adapt to your expectations here. I know everything you wrote is well meant, so thanks for your guidance.
* I think, a web browser is not the best tool for reviewing code. IMO, it is better to see the big picture, actually run the code, maybe use some IDE features to jump around classes and methods methods referenced in the changes in order to see more than changed hunks of diff text, partly out of context. Code is a living creature. I like to watch what it is doing and "get a feeling" for it. Even so, just reading the source code here on GitHub is also a nice readability test. Still, I think that the GitHub web UI alone is inadequate in order to do a proper review, especially when trying to suggest better alternatives to the existing code.
* I also think it makes sense not to focus on the functionality as such a bit more instead of mainly debating style guide issues.

**Comment by kriegaex** [comment]
OK, I pushed another commit and hope that all concerns have been adequately addressed. Like I said, Plexus `JavaVersion` does not cut the cake here, see my lenghier explanation above.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java (hunk 3 above)
Comment (rfscholte) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java]: This is how it could be used:

       // only need for first digit 
       assertThat( JavaVersion.parse( "8.0" ).getValue( 1 ), is( "8" ) );
       // compare with the major version to support 1.x values
       assertThat( JavaVersion.parse( "1.8" ).asMajor().isBefore( JavaVersion.parse( "1.9" ).asMajor() ), is( true ) );       
       assertThat( JavaVersion.parse( "1.8" ).asMajor().isBefore( JavaVersion.parse( "9" ).asMajor() ), is( true ) );
       assertThat( JavaVersion.parse( "8" ).asMajor().isBefore( JavaVersion.parse( "1.9" ).asMajor() ), is( true ) );       
       assertThat( JavaVersion.parse( "8" ).asMajor().isBefore( JavaVersion.parse( "9" ).asMajor() ), is( true ) );

Comment (kriegaex) [reply plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java]: Yeah, nice, but how does that help me here?

**Comment by bmarwell** [comment]
LGTM, great PR! 

The only think left is to get rid of the 1.x / x versions, as Robert suggested:
https://github.com/codehaus-plexus/plexus-compiler/pull/137#discussion_r645423550

I think you don't need the map and can just do here:
https://github.com/codehaus-plexus/plexus-compiler/pull/137/files/3626cbb98aeceb241d4296b3e67f5405afa2a53e#diff-510a8a3bd19d49ca71cb48fbf3b4e02fecee74c16aaf71759799d923be50d7b6R632-R643

Somethink like:
`return JavaVersion.parse( version.trim() ).asMajor() + ".0"`

(of course not .0, but you know what I mean).

**Comment by kriegaex** [comment]
@bmarwell, @rfscholte, maybe I am the problem, not explaining something I thought was obvious well enough, but I really think the two of you are not getting the point, maybe because you simply looks at diffs ina browser, like I said before:
  * Using a `Map` is appropriate, because I actually need to **map** (hence the data structure name) parsed versions to AspectJ constants.
  * Parsing the versions is **not** the problem here, because parsing is not mapping.
  * If I would use `JavaVersion.parse( version.trim() ).asMajor() + ".0"` or something similar, I would have to repeat it for each single key-value pair added to the map, i.e. it would just add duplicated boilerplate and make the code unreadable.

Look, this is what I need to map it to:
image

This is not done by simply parsing a string to a data structure which does not help me save any lines of code when trying to map it into what is really needed here. The only way not to use a map here and map directly from the parsed major versions to the corresponding AspectJ constant, would be to redundantly mimic the math to calculate those constants directly in Plexus, pulling internal AspectJ knowledge into Plexus, which is bound to break if for whatever reason AspectJ would be refactored.

Can you see it now? Parsing is not the issue here, mapping is. Parsing would not save any code, it would add more.

**Comment by kriegaex** [comment]
I really do not like this way of conducting a of code review, which I feel is micro-management and therefore quite disheartening. It feels like I am not a fellow developer but some kind of puppet on strings, being told: "Change the code to exactly how I as a reviewer would have coded it, otherwise the merge will be stalled forever." I know, this is not your intention, guys, but this what it boils down to, intentional or not.

So what shall it be? If you are not suggesting to me but actually commanding me to refactor to switch-case, I shall do so. I am going to think twice before contributing more, though. IMO, this change is not complicated. The talking, me having to explain and re-explain each of my conscious decisions and basically defend them like this was some kind of lawsuit, already was significantly more effort than actually creating the code itself. I know that this ration of coding vs. debate is necessary sometimes, but here I think it is way out of proportion. As much as I like my own contribution. Honestly, it is pretty trivial, not rocket science.

**Comment by kriegaex** [comment]
I hope that all reviewers are happy now, having gotten their way. I still prefer a `Map` in order to **map** A to B, instead of replacing it by a control structure like switch-case, even if the latter is shorter in this case (but only if I use compact one-line style `case "10" : return ClassFileConstants.JDK10;`). I also feel bad about adding a new dependency for one line of code (parsing the version), but what the heck.

**Comment by kriegaex** [comment]
I have no idea, why the integration test is now failing, because locally it passes. Unfortunately, I cannot inspect the `plexus-compiler-its/target/it/aspectj-compiler/build.log`. Maybe you can, if you can log into Travis CI and have more access rights than I do.

I am going to set `<streamLogsOnFailures>true</streamLogsOnFailures>` for now, so I can see what is happening in the failing IT. If you guys want me to revert that after debugging and prefer to rather set a system property for the build job, whatever you prefer is fine with me.

**Comment by kriegaex** [comment]
By the way, do you have a plan B for this?
image

June 15th would be today. Maybe it would make sense to migrate to GitHub Actions. @bmarwell, maybe you like to set up a build job similar to AspectJ Maven. If I would feel a little less frustrated by the code review process just now, I would create another PR myself, but I think I need a little break. I just want to get this one off the table first.

**Comment by kriegaex** [comment]
OK, great, pulling in the new `plexus-java` dependency for the sake of version parsing is causing the integration test to fail:

https://travis-ci.org/github/codehaus-plexus/plexus-compiler/jobs/774585596#L2966:
```text
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project aspectj-compiler:
Execution default-compile of goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile failed:
An API incompatibility was encountered while executing org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile: 
java.lang.NoSuchMethodError: org.codehaus.plexus.languages.java.version.JavaVersion.asMajor()Lorg/codehaus/plexus/languages/java/version/JavaVersion;
```

*Sigh* The ITs use Maven Compiler 3.8.1, which seems to depend on `plexus-java` 0.9.10, which does not have the `asMajor` method yet. So what do we do now? Create a PR for Maven Compiler? Override the version in the AspectJ IT? Stop using the method, doing a bit more parsing by ourselves? This is all so unnecessary.

**Comment by kriegaex** [comment]
OK, I fixed the IT by kicking out `plexus-java` again, but retaining the switch-case which seems to be what you want to have instead of a map.

Hoping that now finally I am permitted to stop chasing my own tail, would you like me to squash any commits and force-push in order to have fewer of them in the upstream commit history?

**Review by rfscholte — APPROVED** [review]

**Comment by kriegaex** [comment]
FYI, I squashed some commits and amended one. Please note my two new comments.

File: plexus-compiler-its/pom.xml [hunk 9]
```diff
@@ -39,7 +39,8 @@
               <goal>verify</goal>
             </goals>
             <configuration>
-              <debug>true</debug>
+              <debug>false</debug>
```
Comment (kriegaex) [inline plexus-compiler-its/pom.xml:42]: When searching for the problem in the IT, I noticed that the debug output was confusing me more than helping me. I think, it is better to activate it on a case-by-case basis. The standard build also does not run in Maven debug mode.

File: plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java [hunk 10]
```diff
@@ -557,83 +564,157 @@ private void checkForAspectJRT( List<String> cp )
...
+            case "9" : return ClassFileConstants.JDK9;
+            case "10" : return ClassFileConstants.JDK10;
+            case "11" : return ClassFileConstants.JDK11;
+            case "12" : return ClassFileConstants.JDK12;
+            case "13" : return ClassFileConstants.JDK13;
+            case "14" : return ClassFileConstants.JDK14;
+            case "15" : return ClassFileConstants.JDK15;
+            case "16" : return ClassFileConstants.JDK16;
+        }
+        throw new CompilerException( "Unknown Java source/target version number: " + version );
```
Comment (kriegaex) [inline plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java:628]: I changed the error message in order to reflect that unlike before, here it is not always about a source version but could also be a target version (or compliance level, more generally).

**Comment by bmarwell** [comment]
Does it make sense to add Java 17 already (if ClassFileConstants already supports it, didn't check)?

**Comment by kriegaex** [comment]
> Does it make sense to add Java 17 already (if ClassFileConstants already supports it, didn't check)?

Did you see any in my screenshot in https://github.com/codehaus-plexus/plexus-compiler/pull/137#issuecomment-861084352? 😉

Thanks for double-checking. Last time I merged JDT Core into AspectJ around the time of the 2021-03 release, which had fresh Java 16 support, no Java 17 yet. It would not be final anyway, even if I merged today, so it is not worth the effort. Those merges are quite tricky.

**Review by bmarwell — APPROVED** [review]

**Comment by kriegaex** [comment]
Thanks for the merge. However, I am not so sure it was a good idea to squash the 4 commits which remained after I squashed by topic myself into a single one. Now there is no spearation of concerns anymore. Debugging or partially reverting something does not get easier that way. In contrast, each single trivial Dependabot version bump is a commit of its own.

**Comment by olamy** [comment]
in this case it;s better to have 4 pull requests if there is different concerns in each commit.

**Comment by kriegaex** [comment]
They are related, but still distinct from one another. 4 separate PRs means more CI builds, more review processes, this PR not being able to depend on the other commits. No problem, Olivier. Probably we just have different philosophies or come from different "schools" of development. I am the agile guy and a proponent of small commits. A long commit history is not a problem, it is actually helpful IMO. But I am content with finally having this PR off the table. so thank you very much again.
